katcp/test/test_testutils.py

```diff
@@ -202,3 +203,26 @@ def test_ioloop_hogging(self):
         DUT(1)
         self.io_loop.add_callback(DUT, 2)
         yield DUT.assert_wait_call_count(2, timeout=0.1)
+
+
+class test_AssertNoMemoryLeaks(unittest.TestCase):
+
+    def test_non_leaky_object(self):
+
+        class NonLeaky(object):
+            pass
+
+        with testutils.assert_no_memory_leaks():
+            obj = NonLeaky()
+            obj = None  # noqa: F841
+
+    def test_leaky_object(self):
+
+        class Leaky(dict):
+            def __init__(self):
+                self.__dict__ = self
+
+        with self.assertRaises(AssertionError):
+            with testutils.assert_no_memory_leaks():
+                obj = Leaky()
+                obj = None  # noqa: F841
```

Comment:
It might be a good idea to have an object that has cyclic references but is collectable e.g.
```python
class Cyclic(object):
    def __init__(self):
        self.me = self
```
and test it one way or the other depending on whether you want to allow this.

Reply:
This one isn't collectable (at least not in Python 2.7 and 3.5).  Should it to be?

Reply:
It seems collectable to me, in both 2.7 and 3.6:
```python3
import gc

class Cyclic(object):
    def __init__(self):
        self.me = self


Cyclic()
gc.collect()
print(gc.garbage)
```

prints an empty list.

Reply:
@bmerry OK - I was using the `gc.DEBUG_SAVEALL` flag, but we shouldn't.  The objects are being freed.  Back to the drawing board.